flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java

```diff
@@ -201,4 +195,32 @@ private static void connectVerticesToConsumedPartitions(
 			}
 		}
 	}
+
+	/**
+	 * Co-location constraints are only used for iteration head and tail.
+	 * A paired head and tail needs to be in the same pipelined region so
+	 * that they can be restarted together.
+	 */
+	private void ensureCoLocatedVerticesInSameRegion() {
+		final Map<CoLocationConstraint, DefaultSchedulingPipelinedRegion> constraintToRegion = new IdentityHashMap<>();
```

Comment:
This seems to also put `CoLocationConstraint == null` into the map.
The test also passes because of it. Is it expected? The test graph (2 disjoint JVs w/o `CoLocationConstraints`) seems to be a normal graph. Why should the test fail?

Reply:
Nullable constraint is a bug here and have resulted in lots of test failures. I have fixed it.
The test was incomplete but I forget about it because it passed due to the nullable constraint bug. I will fix it.

Reply:
done.